sys-utils/nsenter.c

```diff
@@ -557,25 +660,31 @@ int main(int argc, char *argv[])
 #endif
 
 	if (do_all) {
-		if (!namespace_target_pid)
-			errx(EXIT_FAILURE, _("no target PID specified for --all"));
-		for (nsfile = namespace_files; nsfile->nstype; nsfile++) {
-			if (nsfile->fd >= 0)
-				continue;	/* namespace already specified */
-
-			if (!is_usable_namespace(namespace_target_pid, nsfile))
+		struct namespace_file *n = NULL;
+		while ((n = next_nsfile(n, 0))) {
+			if (n->enabled || !is_usable_namespace(namespace_target_pid, n))
 				continue;
-
-			namespaces |= nsfile->nstype;
+			enable_nsfile(n, NULL);
 		}
 	}
 
 	/*
 	 * Open remaining namespace and directory descriptors.
 	 */
-	for (nsfile = namespace_files; nsfile->nstype; nsfile++)
-		if (nsfile->nstype & namespaces)
-			open_namespace_fd(nsfile->nstype, NULL);
+	namespaces = get_namespaces_without_fd();
+	if (namespaces || sock_fd >= 0 || do_user_parent) {
+		if (!namespace_target_pid)
+			errx(EXIT_FAILURE, _("no target PID specified"));
+
+		pid_fd = pidfd_open(namespace_target_pid, 0);
+		if (pid_fd < 0) {
+			if (sock_fd >= 0)
+				err(EXIT_FAILURE, _("failed to pidfd_open() for %d"), namespace_target_pid);
+			if (namespaces)
+				open_namespaces(namespaces);	/* fallback */
```

Comment:
Between Linux 5.3 and 5.7 inclusive, `pidfd_open` works but `setns` does not support pidfds. As `open_namespaces` is only called when `pidfd_open` fails, seems this breaks nsenter on those versions?

Reply:
Ah, I forgot about this kernel trap. I have submitted a new PR to address the issue: https://github.com/util-linux/util-linux/pull/3269. It checks for the kernel version, and while it may not be the most elegant solution, it should be good enough for this case.